### PR TITLE
wip: use read_uncommitted isolation when determining progress

### PR DESCRIPTION
**Pull request by sploiselle**
### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage/src/sink/kafka/util.rs [hunk 1]
```diff
@@ -389,12 +391,19 @@ pub struct ProgressRecord {
...
     progress_topic: String,
     progress_key: String,
     progress_client: Arc<BaseConsumer<TunnelingClientContext<MzClientContext>>>,
+    producer: &KafkaTxProducer,
 ) -> Result<Option<Timestamp>, anyhow::Error> {
+    // Ensure that the producer has initialized transactions and is the current
+    // holder of its transaction ID. We need this to ensure that we get the most
+    // recent offsets from the progress topic.
+    producer.begin_transaction().await?;
+    producer.abort_transaction().await?;
```
Comment (benesch) [inline src/storage/src/sink/kafka/util.rs:405]: Wait, I thought `init_transactions()` was what fenced out old versions of the producer?

Comment (sploiselle) [reply src/storage/src/sink/kafka/util.rs:405]: This is ensuring that `init_transactions` was called; will adjust the note here

Comment (benesch) [reply src/storage/src/sink/kafka/util.rs:405]: Can we enforce that through types instead? It seems like `KafkaTxProducer` could plausibly be the place where we uphold that invariant. Like, stick that type in its own module so the fields are private, and then force construction through an `async fn new() -> Result<..., _>` that forces a call to `init_transaction()` before returning.

Comment (sploiselle) [reply src/storage/src/sink/kafka/util.rs:405]: Sure thing

**Comment by sploiselle** [comment]
Closing in favor of #23380; some of the commits in this PR should be resurrected after https://github.com/MaterializeInc/materialize/pull/23224 merges, namely those that move code out of `storage-client/src/lib` and into `storage/src/sink/kafka

**Pull request closed without merging**